pillarbox-player/src/main/java/ch/srgssr/pillarbox/player/PillarboxMediaMetaDataTracker.kt

```diff
@@ -0,0 +1,144 @@
+            }
+        }
+
+        mediaMetadata.credits?.let {
+            if (currentCreditTracker?.timeRanges != it) {
+                currentCreditTracker?.clear()
+                currentCreditTracker = Tracker(player = player, timeRanges = it, callback = callback)
+            }
+        }
+    }
+
+    private class Tracker<T : TimeRange>(
+        private val player: PillarboxExoPlayer,
+        val timeRanges: List<T>,
+        private val callback: (T?) -> Unit,
+    ) {
+        private val messages: List<PlayerMessage> = createMessages()
+
+        private var currentTimeRange: T? = null
+            set(value) {
+                if (field != value) {
+                    callback(value)
+                    field = value
+                }
+            }
+
+        init {
+            currentTimeRange = timeRanges.firstOrNullAtPosition(player.currentPosition)
+        }
+
+        fun setCurrentPosition(currentPosition: Long) {
+            val currentTimeRange = currentTimeRange
+                ?.takeIf { timeRange -> currentPosition in timeRange }
+                ?: timeRanges.firstOrNullAtPosition(currentPosition)
+            this.currentTimeRange = currentTimeRange
+        }
+
+        private fun createMessages(): List<PlayerMessage> {
+            val messageHandler = PlayerMessage.Target { messageType, message ->
+                @Suppress("UNCHECKED_CAST")
+                val timeRange = message as? T ?: return@Target
+                when (messageType) {
+                    TYPE_ENTER -> currentTimeRange = timeRange
+                    TYPE_EXIT -> {
+                        val nextTimeRange = timeRanges.firstOrNullAtPosition(player.currentPosition)
+                        if (nextTimeRange == null) currentTimeRange = null
+                    }
+                }
+            }
+            val playerMessages = mutableListOf<PlayerMessage>()
+            timeRanges.forEach { timeRange ->
+                val messageEnter = player.createMessage(messageHandler).apply {
+                    deleteAfterDelivery = false
+                    looper = player.applicationLooper
+                    payload = timeRange
+                    setPosition(timeRange.start)
+                    type = TYPE_ENTER
+                    send()
+                }
+                val messageExit = player.createMessage(messageHandler).apply {
+                    deleteAfterDelivery = false
+                    looper = player.applicationLooper
+                    payload = timeRange
+                    setPosition(timeRange.end)
+                    type = TYPE_EXIT
+                    send()
+                }
+                playerMessages.add(messageEnter)
+                playerMessages.add(messageExit)
+            }
+            return playerMessages
+        }
+
+        fun clear() {
+            messages.forEach { it.cancel() }
```

Comment:
Shouldn't we clear the list of messages too?

Reply:
We can't, messages is a immutable list and they are recreated when needed.

Reply:
🤦🏻‍♂️ silly me!